src/analysis/assert_without_msg.d

```diff
@@ -0,0 +1,186 @@
+// Distributed under the Boost Software License, Version 1.0.
+//    (See accompanying file LICENSE_1_0.txt or copy at
+//          http://www.boost.org/LICENSE_1_0.txt)
+
+module analysis.assert_without_msg;
+
+import analysis.base : BaseAnalyzer;
+import dsymbol.scope_ : Scope;
+import dparse.lexer;
+import dparse.ast;
+
+import std.stdio;
+import std.algorithm;
+
+/**
+ * Check that all asserts have an explanatory message.
+ */
+class AssertWithoutMessageCheck : BaseAnalyzer
+{
+	enum string KEY = "dscanner.style.assert_without_msg";
+	enum string MESSAGE = "An assert should have an explanatory message";
+
+	///
+	this(string fileName, const(Scope)* sc, bool skipTests = false)
+	{
+		super(fileName, sc, skipTests);
+	}
+
+	override void visit(const AssertExpression expr)
+	{
+		if (expr.message is null)
+			addErrorMessage(expr.line, expr.column, KEY, MESSAGE);
+		else
+		    preventCheating(expr.message);
+	}
+
+	override void visit(const FunctionCallExpression expr)
+	{
+		if (!isStdExceptionImported)
+			return;
+
+		if (expr.unaryExpression !is null &&
+			expr.unaryExpression.primaryExpression !is null &&
+			expr.unaryExpression.primaryExpression.identifierOrTemplateInstance !is null)
+		{
+			auto ident = expr.unaryExpression.primaryExpression.identifierOrTemplateInstance.identifier;
+			if (ident.text == "enforce" && expr.arguments !is null && expr.arguments.argumentList !is null)
+			{
+				if (expr.arguments.argumentList.items.length < 2)
+				    addErrorMessage(ident.line, ident.column, KEY, MESSAGE);
+                else
+                    preventCheating(expr.arguments.argumentList.items[1]);
+			}
+		}
+	}
+
+	override void visit(const SingleImport sImport)
+	{
+		static immutable stdException = ["std", "exception"];
+		if (sImport.identifierChain.identifiers.map!(a => a.text).equal(stdException))
+			isStdExceptionImported = true;
+	}
+
+	// revert the stack after new scopes
+	override void visit(const Declaration decl)
+	{
+		// be careful - ImportDeclarations don't introduce a new scope
+		if (decl.importDeclaration is null)
+		{
+			bool tmp = isStdExceptionImported;
+			scope(exit) isStdExceptionImported = tmp;
+			decl.accept(this);
+		}
+		else
+			decl.accept(this);
+	}
+
+	mixin ScopedVisit!IfStatement;
+	mixin ScopedVisit!BlockStatement;
+
+	alias visit = BaseAnalyzer.visit;
+
+private:
+	bool isStdExceptionImported;
+
+	template ScopedVisit(NodeType)
+	{
+		override void visit(const NodeType n)
+		{
+			bool tmp = isStdExceptionImported;
+			scope(exit) isStdExceptionImported = tmp;
+			n.accept(this);
+		}
+	}
+
+	void preventCheating(const ExpressionNode exprNode)
```

Comment:
which you would also detect while reviewing Phobos, if this assert thing was really an issue (caugh caugh).